Add radix_sort (#712)
fortran-lang/stdlib: doc/specs/stdlib_sorting.md

low, being of order O(Ln(N)), while the memory requirements of
`ORD_SORT` and `SORT_INDEX` are of order O(N).

#### The `RADIX_SORT` subroutine

> `RADIX_SORT` is a implementation of LSD [radix sort](https://en.wikipedia.org/wiki/Radix_sort),

Reviewer: Do you have another reference than wikipedia (as the algorithm is only briefly described there)?

Author: > Do you have another reference than wikipedia (as the algorithm is only briefly described there)?

Actually I refers to lots of open materials. The following I think best describes the LSD radix sort algorithm:
```suggestion
`RADIX_SORT` is a implementation of LSD [radix sort](https://www.growingwiththeweb.com/sorting/radix-sort-lsd/),
```

> using `256` (one byte) as the radix. It only works for fixed width data,

Reviewer: One byte is 8 bits, right? Why is "one byte" mentioned here?

Author: > One byte is 8 bits, right? Why is "one byte" mentioned here?

Yes, one byte is not 8 bits in some machine, I will remove it.
```suggestion
using `256` as the radix. It only works for fixed width data,
```

thus integers and reals. `RADIX_SORT` is always of O(N) runtime performance
for any input data. For large and random data, it is about five (or more)
times faster than other sort subroutines.

The `RADIX_SORT` needs a buffer that have same size of the input data.

> Your can provide it using `work` arguement, if not the subroutine will

Reviewer: ```suggestion
Your can provide it using `work` argument, if not the subroutine will
```

{!example/sorting/example_sort.f90!}
```

#### `radix_sort` - sorts an input array

##### Status

Experimental

##### Description

Returns an input array with the elements sorted in order of increasing, or
decreasing, value.

##### Syntax

`call [[stdlib_sorting(module):radix_sort(interface)]]( array[, work, reverse] )`

##### Class

Generic subroutine.

##### Arguments

`array` : shall be a rank one array of any of the types:
`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
`real(sp)`, `real(dp)`. It is an `intent(inout)` argument. On return its
input elements will be sorted in order of non-decreasing value.

`work` (optional): shall be a rank one array of the same type as
array, and shall have at least `size(array)` elements. It is an
`intent(inout)` argument, and its contents on return are undefined.

> `reverse` (optional): shall be a scalar of type default logical. It

Reviewer: ```suggestion
`reverse` (optional): shall be a scalar of type default `logical`. It
```

is an `intent(in)` argument. If present with a value of `.true.` then
`array` will be sorted in order of non-increasing values in unstable
order. Otherwise index will sort `array` in order of non-decreasing
values in unstable order.

##### Notes

> `SORT` implements a LSD radix sort algorithm with a `256` radix. For any

Reviewer: ```suggestion
`radix_sort` implements a LSD radix sort algorithm with a `256` radix. For any
```

input data it provides `O(N)` run time performance. If `array` is of

> any type `REAL` the order of its elements on return undefined if any

Reviewer: ```suggestion
any type `real` the order of its elements on return undefined if any
```

  random order.

These benchmarks have been performed on two different compilers, both
on WSL with Ubuntu-20.04, Intel(R) Core(TM) i7-10700 CPU @ 2.9GHz, with

> 32 GB DDR4 memory. The first compiler is GNU Fortran (GCC) 9.4.0, with

Reviewer: It might be good to specify which compiler options/optimizations were used (if non-default ones were used)

Author: > It might be good to specify which compiler options/optimizations were used (if non-default ones were used)

It is built with default CMakeLists.txt
